docs/VirtualMachine.md

```diff
@@ -0,0 +1,160 @@
+Once a trigger is finished it will never fire again.
+
+A data cache uses triggers to know when it can evit and materialize tables.
+
+Triggers fire based on events.
+For example a trigger may fire after receiving an update watermark event.
+See the Flux spec for complete details on events available to triggers.
+
+#### Message bus
+
+The message bus is a queue where trigger events and data messages can be placed.
+Each transformation has its own message bus.
+A transformation consumes messages off its bus and processes them.
+
+### Data flow
+
+With the various data types and operations defined lets be explicit about how data flows through a Flux program.
+
+A Flux program describes a graph through which data flows.
+This graph may have complex data dependencies between transformations.
+Sources produce an inital stream of tables, each table is placed on the message bus for the downstream transformations of the source.
+Each transformation is linked to its downstream transformations via the message bus.
+Sources produces trigger events about the stream which are also placed on the message bus.
+The trigger events cause tables to be materialized from the data cache, if it exists. 
+The materialized tables are placed on the message bus of the downstream transformations.
+If a transformation does not need to make edits to a table it does not need to use the data cache but can simply place its generated tables directly on the downstream message bus.
+
+## Compiler
+
+The compiler is responsible for transforming a semantic graph into an IR.
+Internally the compiler performs several passes in order to produce optimized IR.
+
+1. Initial Pass - Convert the semantic graph into non-optimized IR.
+2. Planner - The planner rewrites the IR to optimize for the structure of the Flux script.
+3. Final Pass - The final pass rewrites the IR for any final optimizations that can be performed.
+
+
+The planner tends to perform more macro optimizations, like merging entire data transformations together.
+The final pass tends to perform more micro optimizations, like rewriting a few instructions to allow for better cache locality.
+
+## Intermediate Representation
+
+The Flux IR consists of a set of instructions and arguments to those instructions.
+The IR is a high level IR, specifically instructions are not required to directly map to hardware operations.
+
+The data types the IR can describe are also high level.
+For example an instruction may take an entire table as its input argument.
+
+The IR describes the both the control flow and the data dependencies between function calls.
+The IR is represented as imperative code.
+As such data dependencies are encoded in the order of the instructions and control flow is encoded through condition jumps to different sections of code.
+
+Here are a few examples of the IR instructions:
+
+* AddInt - Add two integers
+* Stddev - Compute the stddev of a table of data
+* Join - Joins N streams of tables
```

Comment:
So IR instructions can take variable length arguments?

Reply:
I'm not sure complex instructions like `Stddev` and `Join` should be represented as IR instructions. My expectation is that these are `Call` instructions to a runtime library. Thus the VM has an associated "kernel" of standard APIs that we can extend over time.